ingestion/src/kldata.rs

```diff
@@ -106,66 +152,84 @@ fn parse_columns(cols_raw: &str) -> Result<Vec<ObsinnId>, Error> {
         .collect::<Result<Vec<ObsinnId>, Error>>()
 }
 
-fn parse_obs(csv_body: Lines<'_>, columns: &[ObsinnId]) -> Result<Vec<ObsinnObs>, Error> {
+fn parse_obs(
+    csv_body: Lines<'_>,
+    columns: &[ObsinnId],
+    params: Arc<HashMap<String, Param>>,
+) -> Result<Vec<ObsinnObs>, Error> {
     let mut obs = Vec::new();
-    let line_is_empty = || Error::Parse("empty row in kldata csv".to_string());
+    let row_is_empty = || Error::Parse("empty row in kldata csv".to_string());
 
     for row in csv_body {
-        let (timestamp, vals) = {
-            let mut vals = row.split(',');
-
-            let raw_timestamp = vals.next().ok_or_else(line_is_empty)?;
+        let mut vals = row.split(',');
 
-            // TODO: timestamp parsing needs to handle milliseconds and truncated timestamps?
-            let timestamp = NaiveDateTime::parse_from_str(raw_timestamp, "%Y%m%d%H%M%S")
-                .map_err(|e| Error::Parse(e.to_string()))?
-                .and_utc();
+        let raw_timestamp = vals.next().ok_or_else(row_is_empty)?;
 
-            (timestamp, vals)
-        };
+        // TODO: timestamp parsing needs to handle milliseconds and truncated timestamps?
+        let timestamp = NaiveDateTime::parse_from_str(raw_timestamp, "%Y%m%d%H%M%S")
+            .map_err(|e| Error::Parse(e.to_string()))?
+            .and_utc();
 
         for (i, val) in vals.enumerate() {
-            // Should we do some smart bounds-checking??
+            // TODO: should we do some smart bounds-checking??
             let col = columns[i].clone();
 
-            // TODO: parse differently based on param_code?
+            let value = {
+                let param = params.get(&col.param_code).ok_or_else(|| {
+                    Error::Parse(format!("unrecognised param_code '{}'", col.param_code))
+                })?;
```

Comment:
There's maybe a case to be made for just logging this error instead of returning, and treating unknown params like the scalar params (which we're already treating like we don't know their type)